Add support for Gelu operators

gyulaz-htec opened the pull request:
This PR fixes: https://github.com/ROCm/AMDMIGraphX/issues/2758, https://github.com/ROCm/AMDMIGraphX/issues/2754, https://github.com/ROCm/AMDMIGraphX/issues/2757 and adds support for Gelu and it's variants FastGelu and BiasGelu.
Gelu spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Gelu
FastGelu spec: [ContribOperators.md/com.microsoft.FastGelu](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.FastGelu)
BiasGelu spec: [ContribOperators.md/com.microsoft.BiasGelu](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.BiasGelu)

lakhinderwalia commented on src/onnx/parse_gelu.cpp:
```diff
@@ -0,0 +1,116 @@
+    return info.add_instruction(migraphx::make_op("mul"), mul_half, add_one);
+}
+
+instruction_ref parse_gelu_tanh_approx(const onnx_parser::node_info& info, instruction_ref x)
```
Should this function mirror the gelu approximation in https://github.com/ROCm/AMDMIGraphX/pull/2805 ..? Thanks.

gyulaz-htec replied:
Sorry, I didn't notice your comment. This part got removed because of what @pfultz2 mentioned in the [comment](https://github.com/ROCm/AMDMIGraphX/pull/2801#discussion_r1500126975) below. I just kept the `erf` version.
Answering your question: yes, this should mirror the approximation from the link you provided.

pfultz2 commented on src/onnx/parse_gelu.cpp:
```diff
@@ -0,0 +1,116 @@
+        }
+        else
+        {
+            return parse_gelu_no_approx(info, x);
```
We should just insert the `erf` version even if it sets to `tanh`. `rewrite_gelu` will rewrite it to use tanh as needed.

lakhinderwalia replied:
Can't this Gelu be inserted just as a 'gelu' operator, and then rewrite_gelu could insert the appropriate code..? Thanks.

pfultz2 replied:
We should avoid creating operators that are just a composite of other operators. It easy enough to match the erf version(and we already have a matcher for it).

gyulaz-htec commented on src/onnx/parse_gelu.cpp:
```diff
@@ -0,0 +1,76 @@
+        auto x_type             = x->get_shape().type();
+        if(not is_type_float(x_type))
+        {
+            MIGRAPHX_THROW("PAERSE_GELU: input tensor is not a floating type");
```
typo: PAERSE->PARSE

lakhinderwalia commented on test/onnx/verify/gelu_default_half_test.cpp:
```diff
@@ -0,0 +1,59 @@
+    std::vector<float> result_vector;
+    result.visit([&](auto output) { result_vector.assign(output.begin(), output.end()); });
+    tmp = {0.0,
+           -2.3939184e-13,
```
How are these golden values generated, please?

gyulaz-htec replied:
These are cerated according to the `numpy` examples from the spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#examples-59

I've added comments above the gold values how they are generated using numpy.

lakhinderwalia commented on test/onnx/verify/gelu_tanh_test.cpp:
```diff
@@ -0,0 +1,49 @@
+    std::vector<float> result_vector;
+    result.visit([&](auto output) { result_vector.assign(output.begin(), output.end()); });
+    std::vector<float> gold = {
+        0.0f, 0.0f, -1.5497207e-07, -0.15880799f, 0.0, 1.3995717f, 4.8999996f, 8.1999998f, 1000.0f};
```
I suggest that we compare against the reference values coming off the erf() version. Reason is, we are trying to emulate that erf-based function. Only that, here we should lower our accuracy expectations when we use the _approximate_ option. Thanks.

gyulaz-htec replied:
I don't think we should change the gold values to the `erf` version, the end-user would expect to get the `tanh` values when enabling that _approximation_. Also the test passes with these values and these are according to the specification.

lakhinderwalia commented on src/onnx/parse_gelu.cpp:
```diff
@@ -0,0 +1,71 @@
+inline namespace MIGRAPHX_INLINE_NS {
+namespace onnx {
+
+struct parse_gelu : op_parser<parse_gelu>
```
@CharlieL7, should this code mirror the code in find_gelu_erf in rewrite_gelu.cpp? Thanks.

CharlieL7 commented on src/onnx/parse_gelu.cpp:
```diff
@@ -0,0 +1,71 @@
+            MIGRAPHX_THROW("PARSE_GELU: input tensor is not a floating type");
+        }
+
+        // use the erf version for both tanh and erf version
+        // rewrite_gelu will rewrite to use tanh as needed
```
No, we should parse the erf -> erf and the approximation to the tanh approximation with the `pow` instruction. This is because the model intent is different. Yes, the rewrite_gelu pass will replace this parsing if fast_math is enabled for either. However, a model using the approximation intends to have the approximation error. A model using the `erf` version does not. @pfultz2 @lakhinderwalia

gyulaz-htec commented:
@pfultz2 wrote
> We should just insert the `erf` version even if it sets to `tanh`. `rewrite_gelu` will rewrite it to use tanh as needed.
@CharlieL7 wrote
> No, we should parse the erf -> erf and the approximation to the tanh approximation with the pow instruction. This is because the model intent is different. 

Which goes against what @pfultz2 commented first. I can wait for the consensus on this topic.

gyulaz-htec commented:
@pfultz2 @CharlieL7 I've updated the PR with the following:
Added tanh approximation back -> rewrite pass will take care of it if fast_math is enabled
Added parsing of BiasGelu and FastGelu in the `parse_gelu.cpp` as well. I will close the other 2 PR related to those operators.
Updated `gelu_tanh` matcher to match `FastGelu` approximation (`tanh( 0.797885 * x + 0.035677 * x ^ 3 )`) and added a test for it in `rewrite_gelu_test`
